### ISSUE TITLE
Application Gateway is NOT Created

### ISSUE THREAD

**Issue by SOFSPEEL** [Azure/terraform-azurerm-aks#223]
With the configuration shown below:

The application gateway is referenced from the cluster, but the gateway never gets created, note that clicking on the shown checkservice-appgw-dev link results in "resource not found" (and yes I waited the requisite 15 minutes to give it time to be created)

![image](https://user-images.githubusercontent.com/1612646/178770375-ee353f23-c203-44fd-aefd-7303eb5d7f66.png)

despite the ARM json for the cluster looking reasonable to me (note that the blah's below are inserted to avoid publishing sensitive info)

```
           "ingressApplicationGateway": {
                "enabled": true,
                "config": {
                    "applicationGatewayName": "checkservice-appgw-dev",
                    "effectiveApplicationGatewayId": "/subscriptions/{blah}/resourceGroups/MC_eu1d-checkservice-rg_checkservice-cluster-dev_eastus/providers/Microsoft.Network/applicationGateways/checkservice-appgw-dev",
                    "subnetCIDR": "10.52.1.0/24"
                },
                "identity": {
                    "resourceId": "/subscriptions/{blah}/resourcegroups/MC_eu1d-checkservice-rg_checkservice-cluster-dev_eastus/providers/Microsoft.ManagedIdentity/userAssignedIdentities/ingressapplicationgateway-checkservice-cluster-dev",
                    "clientId": "{blah}",
                    "objectId": "{blah}"
                }
            },
```

```
module "network" {
  source              = "Azure/network/azurerm"
  resource_group_name = azurerm_resource_group.rg.name
  address_space       = "10.52.0.0/16"
  subnet_prefixes     = ["10.52.0.0/24"]
  subnet_names        = ["subnet1"]
  depends_on          = [azurerm_resource_group.rg]
}

data "azuread_group" "aks_cluster_admins" {
  # display_name = "AKS-cluster-admins" //todo:
  display_name = "MobileQA"
}

module "aks" {
  source                           = "Azure/aks/azurerm"
  resource_group_name              = azurerm_resource_group.rg.name
  # client_id                        = "your-service-principal-client-appid"
  # client_secret                    = "your-service-principal-client-password"
  kubernetes_version               = "1.23.5"
  orchestrator_version             = "1.23.5"
  prefix                           = "${var.service_name}"
  cluster_name                     = "${var.service_name}-cluster-${var.env}"
  network_plugin                   = "azure"
  vnet_subnet_id                   = module.network.vnet_subnets[0]
  os_disk_size_gb                  = 50
  sku_tier                         = "Paid" # defaults to Free
  enable_role_based_access_control = true
  rbac_aad_admin_group_object_ids  = [data.azuread_group.aks_cluster_admins.id]
  rbac_aad_managed                 = true
  private_cluster_enabled          = true # default value
  enable_http_application_routing  = true
  enable_azure_policy              = true
  # enable_open_service_mesh         = true
  enable_auto_scaling              = true
  # enable_host_encryption           = true //todo
  enable_host_encryption = false
  agents_min_count                 = 1
  agents_max_count                 = 2
  agents_count                     = null # Please set `agents_count` `null` while `enable_auto_scaling` is `true` to avoid possible `agents_count` changes.
  agents_max_pods                  = 100
  agents_pool_name                 = "exnodepool"
  agents_availability_zones        = ["1", "2"]
  agents_type                      = "VirtualMachineScaleSets"

  agents_labels = {
    "nodepool" : "defaultnodepool"
  }

  agents_tags = {
    "Agent" : "defaultnodepoolagent"
  }

  enable_ingress_application_gateway      = true
  ingress_application_gateway_name        = "${var.service_name}-appgw-${var.env}"
  ingress_application_gateway_subnet_cidr = "10.52.1.0/24"

  network_policy                 = "azure"
  net_profile_dns_service_ip     = "10.0.0.10"
  net_profile_docker_bridge_cidr = "170.10.0.1/16"
  net_profile_service_cidr       = "10.0.0.0/16"

  depends_on = [module.network]
}


```

Here are the versions being used:

Terraform v1.2.5
```
 Installing hashicorp/helm v2.6.0...
- Installed hashicorp/helm v2.6.0 (signed by HashiCorp)
- Installing hashicorp/azuread v2.26.1...
- Installed hashicorp/azuread v2.26.1 (signed by HashiCorp)
- Installing hashicorp/tls v3.4.0...
- Installed hashicorp/tls v3.4.0 (signed by HashiCorp)
- Installing hashicorp/local v2.2.3...
- Installed hashicorp/local v2.2.3 (signed by HashiCorp)
- Installing hashicorp/azurerm v2.99.0...
- Installed hashicorp/azurerm v2.99.0 (signed by HashiCorp)
- Installing hashicorp/kubernetes v2.12.1...
- Installed hashicorp/kubernetes v2.12.1 (signed by HashiCorp)
```

**Comment by jonorri** [comment]
@SOFSPEEL Just had the same issue. I'm not sure whether this is the same issue as mine, but I see that you have the client_id and client_secret parameters commented out, just like I had. My thought was that the AKS cluster was actually missing permissions to be able to create the application gateway.
My fix was to create a service principal that had permissions to create an application gateway/subnet/... and now the ingress application gateway is properly created on terraform apply. 

As I said, not sure if you just commented those two lines out for display purposes but wanted to help out here with my fix :)

Hope this helps.

**Comment by levya4** [comment]
Hello @jonorri  ,
Thank you for sharing your fix with us.
Any chance you can share more details on the service principal you created? and how you attach it to the AKS nodes?
Did you use TF to create the principal service?
I'm kinda new to Azure :) 
Thanks !!

**Comment by jonorri** [comment]
@levya4 
I didn´t have a need to have the service principal in terraform. To test this out I just created a service principal that was an owner of my subscription for testing. Working on scoping that access down "just a bit" now.
az ad sp create-for-rbac --name aks-service-principal
az role assignment create --assignee app-id-from-previous-call --scope /subscriptions/your-subscription-id --role Owner

Take appId from the first call and set that as client_id and password as client_secret in the azurerm/aks module invocation. Worked for me.

**Comment by zioproto** [comment]
@SOFSPEEL Thanks for for opening this GitHub issue.

Could you please clarify if you are creating a new cluster greenfield, or if this is an upgrade from an older version of the Terraform module ?

Could you please confirm what is your `tfvars` value for `identity_type` ?

Here the details about the variable:
https://github.com/Azure/terraform-azurerm-aks/blob/master/variables.tf#L152

In case this is an upgrade please be aware of this detail:
https://github.com/Azure/terraform-azurerm-aks#variable-user_assigned_identity_id-has-been-renamed

If you let me know about the `identity_type` you are using I can try to reproduce your issue.

Cheers

**Comment by zioproto** [comment]
@SOFSPEEL any update on this ? Can we further help you ? thanks

**Comment by a-z-i-z** [comment]
@SOFSPEEL At first, I was facing the same issue, here is the solution for it.
In order to provision an Application Gateway in a subnet, the subnet NSG should allow inbound ports which is required by Application gateway. This [doc](https://learn.microsoft.com/en-us/azure/application-gateway/configuration-infrastructure#network-security-groups) would help you which ports need to be opened.

Instead of using `ingress_application_gateway_subnet_cidr` parameter, I would suggest you to first provision a subnet with required ports open and then pass that subnet ID to `ingress_application_gateway_subnet_id` in AKS module. Once done, run `terraform apply` and an Application Gateway along with it's Public IP would be created.

**Comment by zioproto** [comment]
@dkossako I understand you are creating an Application Gateway from the Azure Portal because you want to use with your AKS cluster an existing Application Gateway.

Could you please confirm after you create the Application Gateway what are your values for the following variables ?

```
var.ingress_application_gateway_enabled
var.ingress_application_gateway_id
var.ingress_application_gateway_name
var.ingress_application_gateway_subnet_cidr
var.ingress_application_gateway_subnet_id
```
I need a minimal configuration to reproduce the problem.

Thanks

**Comment by dkossako** [comment]
> I understand you are creating an Application Gateway from the Azure Portal because you want to use with your AKS cluster an existing Application Gateway.

I'm going to create Application Gateway with the code, but it is not deployed. Here's module call, as you can see ingress_application_gateway_* params are configured:
```
  resource_group_name = dependency.rg.outputs.name
  node_resource_group = "${dependency.rg.outputs.name}-node"

  #client_id                        = "your-service-principal-client-appid"
  #client_secret                    = "your-service-principal-client-password"
  kubernetes_version   = "1.24.9"
  orchestrator_version = "1.24.9"
  cluster_name         = "${local.env}-aks"
  prefix               = "${local.account_vars.locals.account_name}"

  network_plugin  = "azure"
  vnet_subnet_id  = dependency.vnet.outputs.vnet_subnets[1]
  os_disk_size_gb = 50
  sku_tier        = "Paid"

  #
  # TODO: There is no "data" resource...
  #
  #rbac_aad_admin_group_object_ids  = [data.azuread_group.aks_cluster_admins.id]
  rbac_aad                          = false
  rbac_aad_managed                  = true
  role_based_access_control_enabled = false
  private_cluster_enabled = false

  # Ingress controller
  ingress_application_gateway_enabled     = true
  ingress_application_gateway_name        = "${local.env}-aks-agw"
  ingress_application_gateway_subnet_cidr = "10.0.10.0/24"

  # default values
  enable_http_application_routing     = true
  enable_azure_policy                 = true
  enable_auto_scaling                 = true
  enable_host_encryption              = false

  # Please set `agents_count` `null` while `enable_auto_scaling` is `true` to avoid possible `agents_count` changes.
  agents_min_count                    = 1
  agents_max_count                    = 2
  agents_count                        = null
  agents_max_pods                     = 33				# Same number of reservations will be made in subnet 
  agents_size                         = "standard_d2as_v5"
  agents_pool_name                    = "exnodepool"
  agents_availability_zones           = ["1", "2"]
  agents_type                         = "VirtualMachineScaleSets"

  agents_labels = {
    "nodepool" : "defaultnodepool"
  }

  agents_tags = {
    "Agent" : "defaultnodepoolagent"
  }

  network_policy                 = "azure"
  net_profile_dns_service_ip     = "10.0.128.10"
  net_profile_docker_bridge_cidr = "170.10.0.1/16"
  net_profile_service_cidr       = "10.0.128.0/17"
}
```

Behavior: AGIC is enabled, but never deployed. Message about 15 minutes is present in my case for last 2 weeks:
![image](https://user-images.githubusercontent.com/11178791/224668105-3f0611ca-11c8-44eb-9f15-1888092a88b5.png)

**Comment by zioproto** [comment]
Maybe I understand what is the problem now.

I see you are passing the subnet "10.0.10.0/24".

Have you created this subnet in the vnet ?

You see that you can pass either `subnet_cidr` or `subnet_id`.
https://github.com/Azure/terraform-azurerm-aks/blob/05a9a3c42847c495262565e8aed3267b67120b1e/main.tf#L234-L243

This assumes that the subnet is already created if you pass it to the API.

**Comment by dkossako** [comment]
It is not created in vnet. As far as I correctly understand documentation - if subnet_cidr is provided then new dedicated subnet will be created with provided CIDR block. If subnet_id is provided - then already created subnet will be used. And that makes sense.

When I'm going to provide existing subnet_cidr then getting error: 
```
Subnet Prefix '10.0.10.0/24' specified for IngressApplicationGateway addon cannot overlap with an existing subnet '10.0.8.0/21' in the AKS Agent Pool's Virtual Network.
```

**Comment by zioproto** [comment]
The subnets must not be overlapping. If you are using already '10.0.8.0/21' in your vnet then '10.0.10.0/24'  is not a valid range.

I dont see any code to create the subnet with that specific value in the provider:
https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/containers/kubernetes_addons.go

**Comment by parkerjgit** [comment]
same issue.
* using terraform azurerm (3.42.0)
* I first provision the subnet (with no NSG), then provision AKS with the subnet ID passed to `ingress_application_gateway_subnet_id`:

```
resource "azurerm_kubernetes_cluster" "aks" {
  ...

  ingress_application_gateway {
    subnet_id = <subnet_id>
  }
}
```

* AKS has a user-assign managed identity associated with it that has "Contributor" role assignment on the subnet scope:

AKS is created successfully, but app gateway is never created:

![image](https://github.com/Azure/terraform-azurerm-aks/assets/1270665/0a917fca-b064-419e-8401-6ecd8d38e225)

**Comment by zioproto** [comment]
@parkerjgit are you using directly the resource `azurerm_kubernetes_cluster` or the module [terraform-azurerm-aks](https://github.com/Azure/terraform-azurerm-aks) ?

Can you please provide a complete code example to reproduce ?

Thanks

**Comment by josedev-union** [comment]
If you didn't create AG manually and just provided the subnet_cidr and gateway_name, AGIC is responsible to deploy a new AG. i.e. the lifecycle of the Application gateway is managed by AGIC.

https://github.com/Azure/application-gateway-kubernetes-ingress/blob/7672d686053cb2c9d4b032d190ac3d6f44fab76b/cmd/appgw-ingress/main.go#L157-L161

The AG deployment could be failed because of some reason in your AGIC. You can check the AGIC log to get the reason why it failed.

For my case, AG creation was failed because of the permission lack.
A managed identity `ingressapplicationgateway-AKS_NAME` is created in the auto-generated resource group(AG will be created in this resource group also) and that is used in AGIC to authenticate azure. But this managed identity doesn't have enough permissions to create AG. From the AGIC log, i got the following error log
```sh
E0630 13:38:01.696441       1 client.go:357] Error while getting virtual network 'default': network.VirtualNetworksClient#Get: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailed" Message="The client 'xxx'(client id of the managed identity `ingressapplicationgateway-AKS_NAME`) with object id 'xxx' does not have authorization to perform action 'Microsoft.Network/virtualNetworks/read' over scope '/subscriptions/yyy/providers/Microsoft.Network/virtualNetworks/default'(this is the resource group where aks cluster is deployed) or the scope is invalid. If access was recently granted, please refresh your credentials."
``` 
I checked the role assignment of the managed identity and found that it is granted no role.
```sh
$ az role assignment list --assignee xxxx
[]
```

I suggest to grant`Network Contributor` role to this managed identity at the scope of `/subscriptions/yyy/providers/Microsoft.Network/virtualNetworks/default`. Ofc, the vnet name will be different and it should be the network where AG will be deployed.

**Comment by avekrivoy** [comment]
> If you didn't create AG manually and just provided the subnet_cidr and gateway_name, AGIC is responsible to deploy a new AG. i.e. the lifecycle of the Application gateway is managed by AGIC.
> 
> https://github.com/Azure/application-gateway-kubernetes-ingress/blob/7672d686053cb2c9d4b032d190ac3d6f44fab76b/cmd/appgw-ingress/main.go#L157-L161
> 
> The AG deployment could be failed because of some reason in your AGIC. You can check the AGIC log to get the reason why it failed.
> 
> For my case, AG creation was failed because of the permission lack. A managed identity `ingressapplicationgateway-AKS_NAME` is created in the auto-generated resource group(AG will be created in this resource group also) and that is used in AGIC to authenticate azure. But this managed identity doesn't have enough permissions to create AG. From the AGIC log, i got the following error log
> 
> ```shell
> E0630 13:38:01.696441       1 client.go:357] Error while getting virtual network 'default': network.VirtualNetworksClient#Get: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="AuthorizationFailed" Message="The client 'xxx'(client id of the managed identity `ingressapplicationgateway-AKS_NAME`) with object id 'xxx' does not have authorization to perform action 'Microsoft.Network/virtualNetworks/read' over scope '/subscriptions/yyy/providers/Microsoft.Network/virtualNetworks/default'(this is the resource group where aks cluster is deployed) or the scope is invalid. If access was recently granted, please refresh your credentials."
> ```
> 
> I checked the role assignment of the managed identity and found that it is granted no role.
> 
> ```shell
> $ az role assignment list --assignee xxxx
> []
> ```
> 
> I suggest to grant`Network Contributor` role to this managed identity at the scope of `/subscriptions/yyy/providers/Microsoft.Network/virtualNetworks/default`. Ofc, the vnet name will be different and it should be the network where AG will be deployed.

@zioproto I'm having the same issue. I assume, permissions on subnet should be granted to the auto-generated MI of the AG.

**Comment by zioproto** [comment]
@avekrivoy you can use the module variable `network_contributor_role_assigned_subnet_ids` and the module will create the necessary role assigments

**Comment by avekrivoy** [comment]
> network_contributor_role_assigned_subnet_ids

It won't work. There's no AG MI object id in the role assignment
https://github.com/Azure/terraform-azurerm-aks/blob/0f07682a7eda030b13f392d31b0350582d8cb145/main.tf#L800

**Comment by avekrivoy** [comment]
I had to add this resource outside of the module, but in fact it should be in

```
resource "azurerm_role_assignment" "application_gateway_subnet_network_contributor" {
  principal_id         = module.aks-dev-cluster[0].ingress_application_gateway.ingress_application_gateway_identity[0].object_id
  scope                = local.application_gateway_subnet_id
  role_definition_name = "Network Contributor"
}
```

**Comment by avekrivoy** [comment]
@zioproto are there any plans to add this to the module?

> I had to add this resource outside of the module, but in fact it should be inside
> 
> ```
> resource "azurerm_role_assignment" "application_gateway_subnet_network_contributor" {
>   principal_id         = module.aks-dev-cluster[0].ingress_application_gateway.ingress_application_gateway_identity[0].object_id
>   scope                = local.application_gateway_subnet_id
>   role_definition_name = "Network Contributor"
> }
> ```

**Comment by zioproto** [comment]
The proposed solution is:

We have to check that `ingress_application_gateway_enabled` is True, and `ingress_application_gateway_subnet_id` is not null, in this case we have to create a new `azurerm_role_assignment` resource in the module.

@avekrivoy are you interested in proposing a PR for the module that I will review ?

**Comment by avekrivoy** [comment]
> ingress_application_gateway_subnet_id

Unfortunately, I don't have time. But I guess the solution should look like this
```
resource "azurerm_role_assignment" "application_gateway_subnet_network_contributor" {
  count                = (var.ingress_application_gateway_enabled == true && var.ingress_application_gateway_subnet_id != null) ? 1 : 0

  principal_id         = azurerm_kubernetes_cluster.main.ingress_application_gateway.ingress_application_gateway_identity[0].object_id
  scope                = var.ingress_application_gateway_subnet_id
  role_definition_name = "Network Contributor"
}
```

**Comment by kandreasyan-korio** [comment]
Try this for SystemAssigned identity.

```  ingress_application_gateway_name      = format("%s-%s", local.name_prefix, "aks-agw")
  ingress_application_gateway_subnet_id = lookup(module.vnet.vnet_subnets_name_id, "aks-agw-snet")
  network_contributor_role_assigned_subnet_ids = {
    aks-agw-snet = lookup(module.vnet.vnet_subnets_name_id, "aks-agw-snet")
  }

**Comment by alexandrud** [comment]
For me the issue was, when you enable AGIC, it creates a separate principal for it (which is a good thing), but it doesn't actually assign it the Network Contributor role. Fortunately the module has the principal ID in the outputs so you can just add that assignment along with your module instance.
`
module "aks" {
  source                                       = "Azure/aks/azurerm"
  version                                      = "~>7.4"
  resource_group_name                          = var.resource_group_name
  node_resource_group                          = "${var.tags.Inventory}-${var.tags.Environment}-${var.tags.Creator}-${var.tags.Role}-nodes"
  os_disk_size_gb                              = var.os_disk_size_gb
  # public_network_access_enabled                = var.public_network_access_enabled
  sku_tier                                     = var.sku_tier
  rbac_aad                                     = var.rbac_aad
  vnet_subnet_id                               = var.vnet_subnet_id
  node_pools                                   = var.nodes
  cluster_name                                 = "${var.tags.Inventory}-${var.tags.Environment}-${var.tags.Creator}-${var.tags.Role}"
  log_analytics_workspace_enabled              = false
  prefix                                       = "${var.tags.Inventory}-${var.tags.Environment}-${var.tags.Creator}"
  role_based_access_control_enabled            = var.role_based_access_control_enabled
  rbac_aad_managed                             = var.rbac_aad_managed
  rbac_aad_admin_group_object_ids              = var.rbac_aad_admin_group_object_ids
  enable_auto_scaling                          = var.enable_auto_scaling
  ingress_application_gateway_enabled          = var.ingress_application_gateway_enabled
  ingress_application_gateway_name             = var.ingress_application_gateway_enabled == false ? null : "${var.tags.Inventory}-${var.tags.Environment}-${var.tags.Creator}-aks-agw"
  ingress_application_gateway_subnet_id        = var.ingress_application_gateway_enabled == false ? null : var.ingress_application_gateway_subnet_id
  network_contributor_role_assigned_subnet_ids = { 
    public = var.ingress_application_gateway_subnet_id
    private = var.vnet_subnet_id
  }
  network_policy                               = var.network_policy
  network_plugin                               = var.network_plugin
}

# This next part is a very ugly workaround for a bug with the azure rm provider
resource "azurerm_role_assignment" "network_contributor_agic" {
  principal_id                     = module.aks.ingress_application_gateway.ingress_application_gateway_identity[0].object_id
  role_definition_name             = "Network Contributor"
  scope                            = var.ingress_application_gateway_subnet_id
}

Hope this helps people!
`

**Comment by Pela2silveira** [comment]
The main problem here is the way Azure deploy AKS in 2 resources groups. The thing is that you usually will expect your apigw to be created in the main resource group, not the MC_.. where the managed identity is created.

So the problem is of permisions between this 2 resource groups and the necesity of app gateway to contribute in the vnet. 

To solve this by terraform contributor role for terraform provider is no enough, so you must use an owner role for terraform asigning role actions.

**Comment by lonegunmanb** [comment]
@Pela2silveira thanks for your reply, would you please give pr #426 a review to see whether this pr could solve the issue? Thanks!

**Comment by zioproto** [comment]
@Pela2silveira @SOFSPEEL the PR #426 should fix this issue completely. Could you folks, or anyone else that is following this issue, please give a try to the latest `main` branch and provider feedback. This is a very long issue, and we don't want to close it too early if anyone is still experiencing problems. Thanks

**Note**: please be aware that the main branch is not stable and might contain breaking changes.

**Comment by lonegunmanb** [comment]
I'm closing this issue since I assume that #426 has fixed this issue. The patch would be released along with `v8.0`. Anyone who has different thoughts or found bugs related to this issue by using the latest `main` branch are welcomed to ping me here, thanks in advance!
